[FLINK-19656] [metrics] filter delimiter in metrics components

echauchot opened the pull request:
## What is the purpose of the change
This PR makes sure the configured per-reporter delimiter is filtered out of the metrics components names.


## Brief change log

- I made the configured reporter delimiter come through to the reporter instance itself. For this I made a default impl in AbstractReporter.open() that loads the configured delimiter and I made all the reporters extend AbstractReporter and I made open() overrides call the default impl. But for configuration name, I chose to copy the conf string "scope.delimiter" which was already defined twice in flink-core/ConfigConstants and flink-core/MetricsOptions rather than making flink-metrics depend on flink-core. Is that correct ?
- In MetricsOptions, the deprecated "metrics.scope.delimiter" conf name was in use. I replaced with the new name "scope.delimiter" in a separate commit. But I find it weird that this conf name is duplicated between MetricsOptions and ConfigConstants
- I made a default impl in AbstractReporter.filterCharacter() that filters out the configured delimiter. I made sure all AbstractReporter subclasses (all the reporters) and overrides call this default impl.

## Verifying this change
Added AbstractReporterITCase test


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no unless MetricOptions is used publicly (I have fixed a deprecation in this file)
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

Added javadocs on AbstractReporter#open and AbstractReporter#filterCharacters to document the default impl and recall users to call super when they override.

echauchot commented:
R: @zentol can you please review this ?
Can you give me your opinion on the comments in the PR ?
before fixing the conflicts (seem due to reformatting on master) I'd like that we agree on the general design of this PR.

echauchot commented:
@flinkbot run azure

zentol commented:
well...this isn't quite the direction I was going for.

The problem I have with it is that we have to modify every single reporter.
This has a few implications:
- this behavior is now essentially opt-in, which ultimately means that user-defined reporters are likely to behave differently, still suffering from the problem at hand
- duplicate logic in reporters
  - which you reduce via inheritance, but:
    - it is now hard-wired to go through the `open()` method, which is essentially legacy code (because new reporters should use factories and pass everything via the constructor, but you can't force that in `AbstractReporter` because legacy reporters may still be around)
    - it is super easy to miss a call to `super#filterCharacters`
    - forces reporters to now pass themselves as the `CharacterFilter`, which is a pattern that we should have never introduced
 - if the new filter actually modifies values then the caching mechanism in the metric groups is rendered pointless (this was admittedly already the case for _some_ reporters, but it is something we want to change in FLINK-14410)

I can just echo my previous statement in the JIRA; we should modify the FrontMetricGroup instead to wrap the given filter (or introduce one where the filter is null) doing the replacement, and add a ( #getAllVariables()) variant that accepts a filter.
Then this behavior will be consistent for every reporter, without having to touch any of them.

echauchot commented:
@zentol thanks for the quick comment. I totally agree with what you said, I was not happy with this solution either. I used this PR as a discussion environment.  This poor design is a consequence of my desire to introduce a default filter for reporters. But, yes, it is better to introduce the default filter in FrontMetricGroup when the provided one is null. I just did not do so because I was afraid that it was not the only place a filter was needed. If you think it is needed only there, fair enough. I'll drop this code and restart from master.

echauchot commented:
@zentol I remade the work from scratch using _FrontMetricGroup_ for filter/delimiter injection:

- I made sure that the configured delimiter is used in filtering when creating the metrics identifier
- I added a _MetricsGroup#getAllVariables()_ version that takes a _CharacterFilter_. And made sure that the configured delimiter is used in filtering variable names and values.
- I added the related tests in _AbstractMetricGroupTest_.

PTAL.

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run travis

echauchot commented:
CI does not seem to be triggered.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java:
```diff
@@ -107,6 +108,8 @@
     /** Flag indicating whether this group has been closed. */
     private volatile boolean closed;
 
+    @VisibleForTesting public static final char DEFAULT_REPLACEMENT = '_';
```
```suggestion
    @VisibleForTesting
    public static final char DEFAULT_REPLACEMENT = '_';
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java:
```diff
@@ -119,12 +122,28 @@ public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
         this.variables = new Map[registry.getNumberReporters() + 1];
     }
 
+    CharacterFilter wrapWithDefaultFilter(CharacterFilter filter, char delimiter) {
```
```suggestion
    private CharacterFilter wrapWithDefaultFilter(CharacterFilter filter, char delimiter) {
```

echauchot replied:
oups I forgot to reduce visibilities, thanks for catching !

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java:
```diff
@@ -119,12 +122,28 @@ public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
     }
 
-    public Map<String, String> getAllVariables(int reporterIndex, Set<String> excludedVariables) {
+    public Map<String, String> getAllVariables(
+            CharacterFilter filter,
+            int reporterIndex,
+            Set<String> excludedVariables,
+            char delimiter) {
```
I'd like to keep the reporterIndex as the first argument because all other arguments are somewhat dependent on it

echauchot replied:
+1

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroupTest.java:
```diff
@@ -415,4 +419,119 @@ public void testGetAllVariablesDoesNotDeadlock() throws InterruptedException {
+        assertThat(allVariables, IsMapContaining.hasValue("v_a"));
+
+        // test default filter, add exclusions to avoid cache hit
+        allVariables = group.getAllVariables(null, 1, Collections.singleton("k*2"), '*');
+        assertThat(allVariables, IsMapContaining.hasKey("k_1"));
+        assertThat(allVariables, IsMapContaining.hasValue("v_1"));
```
There should also be a test that an exclusion that matches the filtered variable (e.g., k_1) is not applied. (i.e., that exclusions are always applied before they are filtered)

That said, I don't think that currently works correctly if a group retrieves variables from a parent, because they will already have the default filter applied to them.

echauchot replied:
as stated in the comment in the code, this exclusion is added only to avoid cache hit. It is not clear enough, I'll replace it by just recreating the group. 
Regarding exclusions, I'll add proper tests for filtered exclusions, thanks for pointing out.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java:
```diff
@@ -137,11 +156,18 @@ public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
+            int cachingIndex,
+            Set<String> excludedVariables,
+            char delimiter) {
+        CharacterFilter newFilter = wrapWithDefaultFilter(filter, delimiter);
```
This implies that variables returned by `parent.getAllVariables()` already had a filter applied to them for a delimiter that may not be matching `delimiter` (i.e., `registry.getDelimiter()`).
Furthermore, this also means that exclusion may not be applied correctly.

I think we need to restrict the filtering to cases where the calls originate from a reporter.

echauchot replied:
ah yeah, thanks for pointing out ! Anyway same goes for the _AbstractMetricGroup#getMetricIdentifier()_ case no ?

zentol replied:
It shouldn't be a problem for `getMetricIdentifier()`; this method only relies on `this.scopeComponents` which is constant and does not fetch anything from parent groups.

Maybe I haven't explained this well.
Let's say the delimiter is `-`, the global default delimiter (returned by `registry.getDelimiter()`) is `.`, and we have a variable defined in a parent group called `-aaa.`.

The result we want is `_aaa.`.

The problem is you retrieved this variable in the first place via `parent.getAllVariables()`. This one doesn't know about your filter, and uses the delimiter returned by `registry.getDelimiter()`. It will apply this filter, resulting in `-aaa_`, and when your filter is applied we end up with `_aaa_`, which is not what we want.

The easiest way I see to solve is to only wrap the filter if the cachingIndexis greater than 0.

echauchot replied:
thanks for the details.
It was clear already for the _getAllVariables()_ inheritance case.

echauchot replied:
besides I don't think that testing `cachingIndex > 0` is enough to detect that the call was done by a reporter because if excluded variables are empty `cachingIndex` is set to 0 (to save memory space) even when the call is done by a reporter.

zentol replied:
Good catch; we have to extend that optimization to only trigger if there is neither a filter nor any excluded variables.

echauchot replied:
what I did for clarity is that `internalGetAllVariables` takes a boolean `calledFromReporter` that is true when the provided `reporterIndex` is valid. `AbstractMetricGroup#getAllVariables()` and `AbstractMetricGroup#getAllVariables(CharacterFilter filter)` that call this method with `registry.getDelimiter()` pass `calledFromReporter == false`. Does it look good to you ?

zentol reviewed (changes requested):
There may be an issue where we are filtering more characters than we should and don't exclude variables that have been retrieved from a parent group.

echauchot commented on flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/MetricGroup.java:
```diff
@@ -186,6 +186,15 @@
+     * @param filter character filter which is applied to the variables and values if not null.
+     * @return map of all variables and their associated value
+     */
+    Map<String, String> getAllVariables(CharacterFilter filter);
```
Yes, I hesitated on that one. Ok I'll revert the interface extension and related changes.

zentol replied:
I think we will need it in the `AbstractMetricGroup`.
Ideally no class works directly against the `FrontMetricGroup`; this currently only happens for some reporters because we do not find the time to properly move `getLogicalScope()` into the `MetricGroup` interface.

echauchot replied:
ok clear, thanks

echauchot commented:
@zentol I updated the PR. PTAL
Please tell me if the added tests are enough in your opinion.
I added different commits to ease your review I'll squash them into isolable commits (most likely into the 4 first ones) when the PR is ok.

zentol commented:
@echauchot I made a final pass over the PR and did some changes:

1) I have removed the filtering of variables. This was my bad (again); filtering the delimiter from variables doesn't make a lot of sense since they are not used in a context where the delimiter usually matters. This could also have resulted in some unexpected behavior from the reporter side, since depending on the delimiter it would no longer be possible to easily look up the value for a specific variable. (image a reporter using '_' as the delimiter; this would change quite a few scope variables).
Filtering variables could be a useful follow-up, but it's a bit tricky due to the lookup issue.
2) Since we only want to do the filtering when used from a reporter I have moved the filter assembly into the `FrontMetricGroup`. I have kept the simplifications in the `AbstractMetricGroup` regarding filters being nullable, by enforcing that they are now not null.
3) I have added a check whether the delimiter used clashes with the default replacement ('_'). If so we use '-' as a backup replacement.

I apologize for wasting your time working to implement the variable filtering.
Thank you for working on this rather nasty part of the codebase.

echauchot commented:
@zentol no problem don't worry. I'm used to iterative thinking / coding. Thanks for the review and the final touches ! 
Indeed, I found this part of the code quite nasty I would have done it simpler but I understand why it is that way : for decoupling the reporters from the rest.

echauchot commented:
@zentol by the way, that is strange to have my name on commits that do not contain the code as I wrote it after rebase but fair enough. 
Maybe next time, adding extra final touches commits would be better.
